File: v1/Python/beginner/05_ShortestPath/telegram.py
# 간단한 다익스트라 문제
# 도시 c에서 보낸 메세지를 받게 되는 도시의 개수와 모두 메세지를 받는 데까지 걸리는 시간

import sys
import heapq


def solve(grp: list[list[tuple[int, int]]], start: int) -> None:
    q = []
    distance = [int(1e9)] * len(grp)

    heapq.heappush(q, (0, start))
    distance[start] = 0

    while q:
        dist, now = heapq.heappop(q)

        if distance[now] < dist:
            continue

        for weight, end_vertex in grp[now]:
            cost = distance[now] + weight

            if distance[end_vertex] > cost:
                distance[end_vertex] = cost
                heapq.heappush(q, (cost, end_vertex))

    cnt = 0
    for dist in distance[1:]:
        if dist != 0 and dist != int(1e9):
            cnt += 1

    print(cnt, end=" ")
    print(max(distance[1:]))


if __name__ == "__main__":
    input = sys.stdin.readline
    n, m, c = map(int, input().split())

    grp = [[] for _ in range(n + 1)]
    for _ in range(m):
        x, y, z = map(int, input().split())
        grp[x].append((z, y))       # x에서 y까지 가중치 z
        grp[y].append((z, x))

    solve(grp, c)

File: v1/Java/boj/Beginner/E_ShortestPath/Telegram.java
package boj.Beginner.E_ShortestPath;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

// Java로 다익스트라 구현하기...

public class Telegram {
    static StringTokenizer st;
    static BufferedReader br =
            new BufferedReader(new InputStreamReader(System.in));

    static List<Node>[] arr;

    static class Node implements Comparable<Node> {
        int end, weight;

        public Node(int end, int weight) {
            this.end = end;
            this.weight = weight;
        }

        @Override
        public int compareTo(Node o) {
            return this.weight - o.weight;
        }
    }

    private static void solve(int start) {
        PriorityQueue<Node> pq = new PriorityQueue<>();
        int[] distance = new int[arr.length];
        Arrays.fill(distance, Integer.MAX_VALUE);

        pq.add(new Node(start, 0));
        distance[start] = 0;

        while (!pq.isEmpty()) {
            Node tmp = pq.poll();
            int now = tmp.end;
            int dist = tmp.weight;

            if (distance[now] < dist)
                continue;

            for (Node node : arr[now]) {
                int end_vtx = node.end;
                int weight = node.weight;
                int cost = distance[now] + weight;

                if (distance[end_vtx] > cost) {
                    distance[end_vtx] = cost;
                    pq.add(new Node(end_vtx, cost));
                }
            }
        }

        int cnt = 0;
        int max_val = -1;
        for (int i = 1; i < distance.length; i++) {
            if (distance[i] != 0 && distance[i] != Integer.MAX_VALUE) {
                cnt++;
                max_val = Math.max(max_val, distance[i]);
            }
        }

        System.out.println(cnt + " " + max_val);
    }

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int c = Integer.parseInt(st.nextToken());

        arr = new ArrayList[n + 1];
        Arrays.setAll(arr, elem -> new ArrayList<>());

        while (m-- > 0) {
            st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            int z = Integer.parseInt(st.nextToken());

            arr[x].add(new Node(y, z));
            arr[y].add(new Node(x, z));
        }

        solve(c);
    }
}
